vdr/store/bbolt.go

```diff
@@ -17,315 +17,321 @@ package store
-	if len(entry.Versions) == 0 {
-		return hash.EmptyHash()
-	}
-
-	return entry.Versions[len(entry.Versions)-1]
-}
-
-var (
-	documentsBucket = []byte("vdrDocuments")
-	versionsBucket  = []byte("vdrDocumentVersions")
+const (
+	// latestBucket has did as key and latest metadata reference as value
+	latestBucket = "latest"
+	// metadataBucket has the metadata reference (did concatenated with version number, starting at 0) as key and the metadataRecord as value
+	metadataBucket = "metadata"
+	// transactionIndexBucket has the transaction reference as key and metadata reference as value
+	transactionIndexBucket = "txRef"
+	// documentsBucket has payload hash as key and document as value
+	documentBucket = "documents"
 )
 
 type bboltStore struct {
 	db *bbolt.DB
 }
 
 // NewBBoltStore returns an instance of a BBolt based VDR store
-func NewBBoltStore(db *bbolt.DB) vdr.Store {
-	return &bboltStore{db: db}
+func NewBBoltStore() vdr.Store {
+	return &bboltStore{}
 }
 
-func (store *bboltStore) storeDocument(tx *bbolt.Tx, document did.Document, metadata vdr.DocumentMetadata) error {
-	documents, err := tx.CreateBucketIfNotExists(documentsBucket)
-	if err != nil {
+func (store *bboltStore) Configure(config core.ServerConfig) error {
+	var err error
+	filePath := path.Join(config.Datadir, "vdr", "didstore.db")
+	if err = os.MkdirAll(path.Join(config.Datadir, "vdr"), os.ModePerm); err != nil {
 		return err
 	}
+	store.db, err = bbolt.Open(filePath, 0600, bbolt.DefaultOptions)
 
-	data, err := json.Marshal(documentVersion{
-		Document: document,
-		Metadata: metadata,
-	})
-	if err != nil {
-		return err
-	}
-
-	if err := documents.Put(metadata.Hash.Slice(), data); err != nil {
-		return err
-	}
+	return err
+}
 
+func (store *bboltStore) Start() error {
+	// already done in Configure
 	return nil
 }
 
-func (store *bboltStore) getDocumentVersion(bucket *bbolt.Bucket, hash hash.SHA256Hash) (*documentVersion, error) {
-	data := bucket.Get(hash.Slice())
-	if data == nil {
-		return nil, nil
-	}
-
-	result := &documentVersion{}
-
-	if err := json.Unmarshal(data, result); err != nil {
-		return nil, err
+func (store *bboltStore) Shutdown() error {
+	if store.db != nil {
+		return store.db.Close()
 	}
+	return nil
+}
 
-	return result, nil
+type metadataRecord struct {
+	Deactivated bool `json:"deactivated"`
+	DID         string
+	Version     int `json:"version"`
+	Metadata    vdr.DocumentMetadata
+	// PrevRecord holds the previous metadataRecord reference (DID + version) as string
+	PrevMetaRef []byte `json:"prevMetaRef"`
 }
 
-// Iterate loops over all the latest versions of the stored DID Documents and applies fn
-func (store *bboltStore) Iterate(fn vdr.DocIterator) error {
-	return store.db.View(func(tx *bbolt.Tx) error {
-		versions := tx.Bucket(versionsBucket)
-		if versions == nil {
-			return nil
-		}
+func (mr metadataRecord) ref() []byte {
+	metaRefString := fmt.Sprintf("%s%06d", mr.DID, mr.Version)
```

Comment:
At time of writing it was a good idea to keep the different versions in order for efficient page fetching from disk. But given the fact that other DID documents are scattered all over disk (based on DID) it doesn't really optimize a lot. It would optimize a search query where a previous DID Document would be fetched. This would only be the case through a manual API call, normal operation would just need the last one. It would be way better to optimize the location of all the latest DID Docs. Better to leave that to an optimization step when dealing with storage in the greater picture.

For now, I'll just remove the `06`